core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -569,28 +569,34 @@ private IndexShardState changeState(IndexShardState newState, String reason) {
         return result;
     }
 
+    public Engine.NoOp prepareMarkingSeqNoAsNoOp(long seqNo, String reason) {
+        verifyReplicationTarget();
+        long startTime = System.nanoTime();
+        return new Engine.NoOp(seqNo, primaryTerm, Engine.Operation.Origin.REPLICA, startTime, reason);
+    }
+
+    public Engine.NoOpResult markSeqNoAsNoOp(Engine.NoOp noOp) throws IOException {
```

Comment:
Most of the other `IndexShard` methods correspond to their `Engine` implementations, so I think this should be called just `noOp`